### PR TITLE
feat(gear-programs): bridging-payment for vara supply

### PR DESCRIPTION
**Pull request by ByteNacked**
- [x]  Add bridging payment for Vara supply contract
- [ ] ~~Refactor bridging payment contracts~~    **In next PR**
- [ ]  ~~Use more recent `sails-rs` project structure in `bridging payment ` contracts~~   **In next PR**

@reviewer-or-team

### CODE REVIEW

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+use utils::maybe_event_or_panic_async;
+use vft_treasury_msg::send_message_to_treasury;
+
+pub struct BridgingPayment<ExecContext> {
+    exec_context: ExecContext,
+}
+
+#[derive(Encode, Decode, TypeInfo)]
+pub enum BridgingPaymentEvents {
+    DepositVaraToTreasury {
```
Comment (mertwole) [inline gear-programs/bridging-payment-vara-supply/src/services/mod.rs:25]:
```suggestion
    DepositToTreasury {
```

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+#[service(events = BridgingPaymentEvents)]
+impl<T> BridgingPayment<T>
+where
+    T: ExecContext,
+{
+    pub fn set_fee(&mut self, fee: u128) {
+        let data = self.data();
+        if data.admin_address != self.exec_context.actor_id() {
+            panic!("Not admin");
+        }
```
Comment (mertwole) [inline gear-programs/bridging-payment-vara-supply/src/services/mod.rs:145]: I suggest moving it to separate fn `ensure_admin` and use it in all methods that require admin access

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+    pub fn reclaim_fee(&mut self) {
+        let data = self.data();
+        if data.admin_address != self.exec_context.actor_id() {
+            panic!("Not admin");
+        }
+        let fee_balance = exec::value_available();
+        msg::send_with_gas(data.admin_address, "", 0, fee_balance).expect("Failed to reclaim fees");
+    }
+
+    pub fn update_vft_treasury_address(&mut self, new_vft_treasury_address: ActorId) {
```
Comment (mertwole) [inline gear-programs/bridging-payment-vara-supply/src/services/mod.rs:159]: I'd remove it as it's easier to re-deploy program if vft treasury address changes

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        if let Some(reply_timeout) = reply_timeout {
+            self.config_mut().reply_timeout = reply_timeout;
+        }
+
+        if let Some(gas_for_request_to_treasury_msg) = gas_for_request_to_treasury_msg {
+            self.config_mut().gas_for_request_to_treasury_msg = gas_for_request_to_treasury_msg;
+        }
+    }
+
+    pub async fn request_transaction(
```
Comment (mertwole) [inline gear-programs/bridging-payment-vara-supply/src/services/mod.rs:199]:
```suggestion
    pub async fn request(
```

File: gear-programs/bridging-payment-vara-supply/src/services/utils.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
+use super::{
+    error::Error,
+    msg_tracker::{msg_tracker_mut, MessageStatus},
+    vft_treasury::vft_treasury::io as vft_treasury_io,
+    vft_treasury::Error as VftTreasuryError,
+};
+use sails_rs::calls::ActionIo;
+use sails_rs::prelude::*;
+
+macro_rules! maybe_event_or_panic_async {
```
Comment (mertwole) [inline gear-programs/bridging-payment-vara-supply/src/services/utils.rs:10]: can you re-write this as fn or remove it? It's used only in 2 places and it's hard to understand this logic outside of the context of `request_tx` or `continue_tx`

File: gear-programs/bridging-payment-vara-supply/src/wasm/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+bridging_payment_vara_supply.workspace = true
+
+[build-dependencies]
+gwasm-builder.workspace = true
+sails-idl-gen.workspace = true
+bridging_payment_vara_supply.workspace = true
+
+[dev-dependencies]
+gtest.workspace = true
+extended_vft_wasm = { git = "https://github.com/gear-foundation/standards/", branch = "gstd-pinned-v1.5.0"}
```
[the remark this answers is not in the document]
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/wasm/Cargo.toml]: Fixed

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs (hunk 1 above)
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/services/mod.rs]: Fixed

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs (hunk 2 above)
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/services/mod.rs]: Fixed

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs (hunk 3 above)
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/services/mod.rs]: Fixed

File: gear-programs/bridging-payment-vara-supply/src/services/mod.rs (hunk 4 above)
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/services/mod.rs]: Fixed

File: gear-programs/bridging-payment-vara-supply/src/services/utils.rs (hunk 5 above)
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/services/utils.rs]: Fixed

File: gear-programs/bridging-payment-vara-supply/src/services/utils.rs [hunk 7]
```diff
@@ -0,0 +1,32 @@
+use super::{
+    error::Error, vft_treasury::vft_treasury::io as vft_treasury_io,
+    vft_treasury::Error as VftTreasuryError,
+};
+use sails_rs::calls::ActionIo;
+use sails_rs::prelude::*;
+
+pub async fn send_message_with_gas_for_reply(
```
Comment (mertwole) [inline gear-programs/bridging-payment-vara-supply/src/services/utils.rs:8]: Maybe also encode message here? anyway it depends on the message format of `vft-treasury` and as I see `send_message_with_gas_for_reply` is called only inside `send_message_to_treasury`

**Review by mertwole — APPROVED** [review]

File: gear-programs/bridging-payment-vara-supply/src/services/utils.rs (hunk 7 above)
Comment (ByteNacked) [reply gear-programs/bridging-payment-vara-supply/src/services/utils.rs:4]: Extracted encoding/decoding  to `send_message_to_treasury`

**Pull request merged**